### PR TITLE
METRON-2123: Expand Stellar JOIN to work on all Iterables

### PR DESCRIPTION
**Pull request by mmiklavc**
## Contributor Comments

https://issues.apache.org/jira/browse/METRON-2123

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- n/a If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- n/a Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

- n/a Have you ensured that any documentation diagrams have been updated, along with their source files, using [draw.io](https://www.draw.io/)? See [Metron Development Guidelines](https://cwiki.apache.org/confluence/display/METRON/Development+Guidelines) for instructions.

### CODE REVIEW

File: metron-stellar/stellar-common/README.md [hunk 1]
```diff
@@ -702,9 +702,9 @@ Where:
   * Returns: True if the string is a valid URL and false if otherwise.
 
 ### `JOIN`
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/README.md]: Do we explicitly want to say `Java Iterable` here?

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java [hunk 2]
```diff
@@ -133,15 +131,15 @@ public Object apply(List<Object> strings) {
   }
 
   @Stellar( name="JOIN"
-          , description="Joins the components in the list of strings with the specified delimiter."
-          , params = { "list - List of strings", "delim - String delimiter"}
+          , description="Joins the non-null items in the iterable as strings with the specified delimiter. Null items are dropped."
+          , params = { "iterable - iterable (e.g. List, LinkedHashSet, etc.) of items treated as strings", "delim - String delimiter"}
           , returns = "String"
           )
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: do we need to SurpressWarnings("unchecked")?

Comment (mmiklavc) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: > do we need to SurpressWarnings("unchecked")?

That's always been there - it's calling out the unchecked type cast `(Iterable<Object>) args.get(0);` (previously `(List<Object>) args.get(0);` Alternatively, we can add extra instanceof checks, but it's really 6 of one, half a dozen of the other. You get a classcastexception.

File: metron-stellar/stellar-common/README.md (hunk 1 above)
Comment (mmiklavc) [reply metron-stellar/stellar-common/README.md]: Seems reasonable, done.

**Comment by mmiklavc** [comment]
### Testing Instructions

This is test via unit tests, however you can also spin up the Stellar CLI and do the following for manual validation:

**Test for regressions on list type**
```
mylist := [ 1, 2, 3 ]
JOIN(mylist, ',')
"1,2,3"
```

**Test set types**
```
JOIN(SET_INIT( [ 'foo', 'bar', 1, 2, 'baz' ] ), ',')
"foo,bar,1,2,baz"
```

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java (hunk 2 above)
Comment (mmiklavc) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: Any thoughts on this @ottobackwards?

Comment (ottobackwards) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: I'm not saying it is new, but we can add the surpress while we are here.

**Comment by ottobackwards** [comment]
+1 by inspection

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java (hunk 2 above)
Comment (mmiklavc) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: I'm sorry, I don't think I follow - line 140 just above the `apply(...)` method already has a "@SuppressWarnings" annotation. Are you asking whether we need to remove it, or were you suggesting adding another one somewhere?

Comment (ottobackwards) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: please, ignore this, sorry.

**Comment by ottobackwards** [comment]
+2

**Comment by nickwallen** [comment]
Wait a tick; we can +2 things??? :)

**Pull request closed without merging**